x-pack/plugins/security_solution/common/detection_engine/build_exceptions_filter.ts

```diff
@@ -0,0 +1,303 @@
+              [field]: value[0],
+            },
+          },
+        ],
+        minimum_should_match: 1,
+      },
+    };
+  }
+
+  return {
+    bool: {
+      should: value.map((val) => {
+        return {
+          bool: {
+            should: [
+              {
+                match_phrase: {
+                  [field]: val,
+                },
+              },
+            ],
+            minimum_should_match: 1,
+          },
+        };
+      }),
+      minimum_should_match: 1,
+    },
+  };
+};
+
+export const buildMatchAnyClause = (entry: EntryMatchAny): BooleanFilter => {
+  const { operator } = entry;
+  const matchAnyClause = getBaseMatchAnyClause(entry);
+
+  if (operator === 'excluded') {
+    return buildExclusionClause(matchAnyClause);
+  } else {
+    return matchAnyClause;
+  }
+};
+
+export const buildExistsClause = (entry: EntryExists): BooleanFilter => {
+  const { field, operator } = entry;
+  const existsClause = {
+    bool: {
+      should: [
+        {
+          exists: {
+            field,
+          },
+        },
+      ],
+      minimum_should_match: 1,
+    },
+  };
+
+  if (operator === 'excluded') {
+    return buildExclusionClause(existsClause);
+  } else {
+    return existsClause;
+  }
+};
+
+const isBooleanFilter = (clause: object): clause is BooleanFilter => {
+  const keys = Object.keys(clause);
+  return keys.includes('bool') != null;
+};
+
+export const getBaseNestedClause = (
+  entries: NonListEntry[],
+  parentField: string
+): BooleanFilter => {
+  if (entries.length === 1) {
+    const [singleNestedEntry] = entries;
+    const innerClause = createInnerAndClauses(singleNestedEntry, parentField);
+    return isBooleanFilter(innerClause) ? innerClause : { bool: {} };
+  }
+
+  return {
+    bool: {
+      filter: entries.map((nestedEntry) => createInnerAndClauses(nestedEntry, parentField)),
+    },
+  };
+};
+
+export const buildNestedClause = (entry: EntryNested): NestedFilter => {
+  const { field, entries } = entry;
+
+  const baseNestedClause = getBaseNestedClause(entries, field);
+
+  return {
+    nested: {
+      path: field,
+      query: baseNestedClause,
+      score_mode: 'none',
+    },
+  };
+};
+
+export const createInnerAndClauses = (
+  entry: NonListEntry,
+  parent?: string
+): BooleanFilter | NestedFilter => {
+  if (entriesExists.is(entry)) {
+    const field = parent != null ? `${parent}.${entry.field}` : entry.field;
+    return buildExistsClause({ ...entry, field });
+  } else if (entriesMatch.is(entry)) {
+    const field = parent != null ? `${parent}.${entry.field}` : entry.field;
+    return buildMatchClause({ ...entry, field });
+  } else if (entriesMatchAny.is(entry)) {
+    const field = parent != null ? `${parent}.${entry.field}` : entry.field;
+    return buildMatchAnyClause({ ...entry, field });
+  } else if (entriesNested.is(entry)) {
+    return buildNestedClause(entry);
+  } else {
+    throw new Error(`Unexpected exception entry: ${entry}`);
```

Comment:
nit: TypeError here?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypeError

 ```suggestion
    throw new TypeError(`Unexpected exception entry: ${entry}`);
```